### ISSUE TITLE
PGP key canonical value does not match input

### ISSUE THREAD

**Issue by Hayden-IO** [sigstore/rekor#1251]
tldr:
* `CanonicalValue` re-encodes the PGP key
* `CanonicalValue` does not add a newline

To recreate:

* Read in an armored PGP key, like here - https://github.com/sigstore/rekor/blob/main/pkg/types/rekord/v0.0.1/entry_test.go#L54
* Call `key := pgp.NewPublicKey(bytes.NewReader(<input>))`
* Compare `key.CanonicalValue()` to the `<input>`, they will differ.

An example set of keys:

Original:

```
-----BEGIN PGP PUBLIC KEY BLOCK-----

mQENBGDX7rcBCACdHDAoYc4ZxrxaVNfkPhHwNjUDKMTkQIC/laU64yUaR4ZyHu9W
5OH1dU6+yCFeysyNpwa0v1701Q6goJkd4BLVNK7JEKTp7gHPznSJkdFsFAN7P7gq
UB+ZJgVRr9g9F83fNJXfzS5DVxC691yn1O5fm9sGlXRYsqhktuhOX8CKgQpdvYNw
aO8gfdK2tB5ebsk7yJUvWj0h12fnQgxRlBtCihN8Hf+tmXVzKZZb27tN0SfdXmhX
aSuHNMUunS/hz3xjc3rXkaWZ4lSqqZOX00C0hSRIey4L6vNBHi75oFzKFCesZchB
VSGs//iPllQMHiVGGyXRuOvoC1BjnDT84WIdABEBAAG0DG5vdEByZWFsLmNvbYkB
VAQTAQgAPhYhBI1o/WP1Lc9+qnJWW8PIKK+2aBwXBQJg1+63AhsDBQkDwmcABQsJ
CAcCBhUKCQgLAgQWAgMBAh4BAheAAAoJEMPIKK+2aBwXlQ8IAJPpkjiACt9aaJIF
Dn5DYVu2WEp96Ft9SwV4Ege2cPKQy8JVtyweSF+lgll0N0r/G3uajmuaAGxGo9x4
iuhRgGeDWheO9TFKSMLb0Fj7FyRg1XOHIXujh7WlnnYVnG3hg8LYhzJenfyNLAsO
6CtER2naadhJMbKwhNl6X86ILtjC+LKyQPNK9ECMVUwZHULzEskoJLsOFW6Y/I2F
q8qoASQK7dmM0ZJHAkwZTeJgAxQvBQtHkTA7LxBXNYXbE4IaMAUWMyL9rnlkUn5e
iAr9W+OAyUPwS8Wn5s/as0YG7uoFvfAYjx5DLsL1Lf3UBu1FXNNvhg30WakTEDYp
kyTO3QG5AQ0EYNfutwEIAMpplZyyHhcI+Mw8MWyKovRqHoaEUeG754XS5c/YEi+T
h55g4BkWf0Tq+ZZ1C0QGdL/h6uzKeD/hwqRJk6tSEdWEV11s6aglS7ps9Z3kady8
7DiwDC7GOPueVRtLfyjQCGfpuoKriDEV155W13BOTGiMin9evWfUTeZ5XFO5UI5S
CHDY8u/zkepVFHE/vM/GX5eo2NTkeZWBwzDvAd0pmhLyCKxfJdLASIKjSmF2aPnu
WCXT732Aj3JY2vL6Z3k1rFT3gbScTFFL+HxytP5QAoR0IwVF/Q5Fw2wKd/1PHt8I
uwDd2eqBrt32JPYSugmeZnEZFDSIYXyui4jFFipf0vEAEQEAAYkBPAQYAQgAJhYh
BI1o/WP1Lc9+qnJWW8PIKK+2aBwXBQJg1+63AhsMBQkDwmcAAAoJEMPIKK+2aBwX
Y6gH/2p6EKEA0XnnIcSae2umq1Xz39IvQHOcvRPfgybtMd2uVqhup17ONhOXkzpa
qxlTF3djDzu4eisK5XpeC/5dHnozHPiey3AqOvVpdPPzvfb8O8pASu+uo+eE0/9G
dww401OjMKi3nVfvLitcttccbszWoMpCGybJvoKBh4bsjSfej1OXorEJNATXazlN
CVUhBNDz8mB+e3Dmz/3vzQryfksi0LxU1YpX1f2sWxMq8kLp3yJWnjRIYJQ9A7Vs
HMv65ebKlfpoUGCtWiKD2bF5lysFoiL+gDNzsbb9ylWfljmYOFw+wNCFvf/yHYxX
xfmrMdZ0nbvRfjuMr27eOY/vO+M=
=IqHu
-----END PGP PUBLIC KEY BLOCK-----
```

One from `CanonicalValue`

```
-----BEGIN PGP PUBLIC KEY BLOCK-----

xsBNBGDX7rcBCACdHDAoYc4ZxrxaVNfkPhHwNjUDKMTkQIC/laU64yUaR4ZyHu9W
5OH1dU6+yCFeysyNpwa0v1701Q6goJkd4BLVNK7JEKTp7gHPznSJkdFsFAN7P7gq
UB+ZJgVRr9g9F83fNJXfzS5DVxC691yn1O5fm9sGlXRYsqhktuhOX8CKgQpdvYNw
aO8gfdK2tB5ebsk7yJUvWj0h12fnQgxRlBtCihN8Hf+tmXVzKZZb27tN0SfdXmhX
aSuHNMUunS/hz3xjc3rXkaWZ4lSqqZOX00C0hSRIey4L6vNBHi75oFzKFCesZchB
VSGs//iPllQMHiVGGyXRuOvoC1BjnDT84WIdABEBAAHNDG5vdEByZWFsLmNvbcLA
lAQTAQgAPhYhBI1o/WP1Lc9+qnJWW8PIKK+2aBwXBQJg1+63AhsDBQkDwmcABQsJ
CAcCBhUKCQgLAgQWAgMBAh4BAheAAAoJEMPIKK+2aBwXlQ8IAJPpkjiACt9aaJIF
Dn5DYVu2WEp96Ft9SwV4Ege2cPKQy8JVtyweSF+lgll0N0r/G3uajmuaAGxGo9x4
iuhRgGeDWheO9TFKSMLb0Fj7FyRg1XOHIXujh7WlnnYVnG3hg8LYhzJenfyNLAsO
6CtER2naadhJMbKwhNl6X86ILtjC+LKyQPNK9ECMVUwZHULzEskoJLsOFW6Y/I2F
q8qoASQK7dmM0ZJHAkwZTeJgAxQvBQtHkTA7LxBXNYXbE4IaMAUWMyL9rnlkUn5e
iAr9W+OAyUPwS8Wn5s/as0YG7uoFvfAYjx5DLsL1Lf3UBu1FXNNvhg30WakTEDYp
kyTO3QHOwE0EYNfutwEIAMpplZyyHhcI+Mw8MWyKovRqHoaEUeG754XS5c/YEi+T
h55g4BkWf0Tq+ZZ1C0QGdL/h6uzKeD/hwqRJk6tSEdWEV11s6aglS7ps9Z3kady8
7DiwDC7GOPueVRtLfyjQCGfpuoKriDEV155W13BOTGiMin9evWfUTeZ5XFO5UI5S
CHDY8u/zkepVFHE/vM/GX5eo2NTkeZWBwzDvAd0pmhLyCKxfJdLASIKjSmF2aPnu
WCXT732Aj3JY2vL6Z3k1rFT3gbScTFFL+HxytP5QAoR0IwVF/Q5Fw2wKd/1PHt8I
uwDd2eqBrt32JPYSugmeZnEZFDSIYXyui4jFFipf0vEAEQEAAcLAfAQYAQgAJhYh
BI1o/WP1Lc9+qnJWW8PIKK+2aBwXBQJg1+63AhsMBQkDwmcAAAoJEMPIKK+2aBwX
Y6gH/2p6EKEA0XnnIcSae2umq1Xz39IvQHOcvRPfgybtMd2uVqhup17ONhOXkzpa
qxlTF3djDzu4eisK5XpeC/5dHnozHPiey3AqOvVpdPPzvfb8O8pASu+uo+eE0/9G
dww401OjMKi3nVfvLitcttccbszWoMpCGybJvoKBh4bsjSfej1OXorEJNATXazlN
CVUhBNDz8mB+e3Dmz/3vzQryfksi0LxU1YpX1f2sWxMq8kLp3yJWnjRIYJQ9A7Vs
HMv65ebKlfpoUGCtWiKD2bF5lysFoiL+gDNzsbb9ylWfljmYOFw+wNCFvf/yHYxX
xfmrMdZ0nbvRfjuMr27eOY/vO+M=
=oMq1
-----END PGP PUBLIC KEY BLOCK-----
```

Something else to note is that `CanonicalValue` drops the newline of the input too.

Looking at the keys:

Original:

```
$ gpg --list-packets origkey.asc
# off=0 ctb=99 tag=6 hlen=3 plen=269
:public key packet:
	version 4, algo 1, created 1624764087, expires 0
	pkey[0]: [2048 bits]
	pkey[1]: [17 bits]
	keyid: C3C828AFB6681C17
# off=272 ctb=b4 tag=13 hlen=2 plen=12
:user ID packet: "not@real.com"
# off=286 ctb=89 tag=2 hlen=3 plen=340
:signature packet: algo 1, keyid C3C828AFB6681C17
	version 4, created 1624764087, md5len 0, sigclass 0x13
	digest algo 8, begin of digest 95 0f
	hashed subpkt 33 len 21 (issuer fpr v4 8D68FD63F52DCF7EAA72565BC3C828AFB6681C17)
	hashed subpkt 2 len 4 (sig created 2021-06-27)
	hashed subpkt 27 len 1 (key flags: 03)
	hashed subpkt 9 len 4 (key expires after 2y0d0h0m)
	hashed subpkt 11 len 4 (pref-sym-algos: 9 8 7 2)
	hashed subpkt 21 len 5 (pref-hash-algos: 10 9 8 11 2)
	hashed subpkt 22 len 3 (pref-zip-algos: 2 3 1)
	hashed subpkt 30 len 1 (features: 01)
	hashed subpkt 23 len 1 (keyserver preferences: 80)
	subpkt 16 len 8 (issuer key ID C3C828AFB6681C17)
	data: [2048 bits]
# off=629 ctb=b9 tag=14 hlen=3 plen=269
:public sub key packet:
	version 4, algo 1, created 1624764087, expires 0
	pkey[0]: [2048 bits]
	pkey[1]: [17 bits]
	keyid: 46EDEC9576D374CA
# off=901 ctb=89 tag=2 hlen=3 plen=316
:signature packet: algo 1, keyid C3C828AFB6681C17
	version 4, created 1624764087, md5len 0, sigclass 0x18
	digest algo 8, begin of digest 63 a8
	hashed subpkt 33 len 21 (issuer fpr v4 8D68FD63F52DCF7EAA72565BC3C828AFB6681C17)
	hashed subpkt 2 len 4 (sig created 2021-06-27)
	hashed subpkt 27 len 1 (key flags: 0C)
	hashed subpkt 9 len 4 (key expires after 2y0d0h0m)
	subpkt 16 len 8 (issuer key ID C3C828AFB6681C17)
	data: [2047 bits]
```

Re-encoded key:

```
$ gpg --list-packets key.asc
# off=0 ctb=c6 tag=6 hlen=3 plen=269 new-ctb
:public key packet:
	version 4, algo 1, created 1624764087, expires 0
-----BEGIN PGP PUBLIC KEY BLOCK-----
	pkey[0]: [2048 bits]
	pkey[1]: [17 bits]
	keyid: C3C828AFB6681C17
# off=272 ctb=cd tag=13 hlen=2 plen=12 new-ctb
:user ID packet: "not@real.com"
# off=286 ctb=c2 tag=2 hlen=3 plen=340 new-ctb
:signature packet: algo 1, keyid C3C828AFB6681C17
	version 4, created 1624764087, md5len 0, sigclass 0x13
	digest algo 8, begin of digest 95 0f
	hashed subpkt 33 len 21 (issuer fpr v4 8D68FD63F52DCF7EAA72565BC3C828AFB6681C17)
	hashed subpkt 2 len 4 (sig created 2021-06-27)
	hashed subpkt 27 len 1 (key flags: 03)
	hashed subpkt 9 len 4 (key expires after 2y0d0h0m)
	hashed subpkt 11 len 4 (pref-sym-algos: 9 8 7 2)
	hashed subpkt 21 len 5 (pref-hash-algos: 10 9 8 11 2)
	hashed subpkt 22 len 3 (pref-zip-algos: 2 3 1)
	hashed subpkt 30 len 1 (features: 01)
	hashed subpkt 23 len 1 (keyserver preferences: 80)
	subpkt 16 len 8 (issuer key ID C3C828AFB6681C17)
	data: [2048 bits]
# off=629 ctb=ce tag=14 hlen=3 plen=269 new-ctb
:public sub key packet:
	version 4, algo 1, created 1624764087, expires 0
	pkey[0]: [2048 bits]
	pkey[1]: [17 bits]
	keyid: 46EDEC9576D374CA
# off=901 ctb=c2 tag=2 hlen=3 plen=316 new-ctb
:signature packet: algo 1, keyid C3C828AFB6681C17
	version 4, created 1624764087, md5len 0, sigclass 0x18
	digest algo 8, begin of digest 63 a8
	hashed subpkt 33 len 21 (issuer fpr v4 8D68FD63F52DCF7EAA72565BC3C828AFB6681C17)
	hashed subpkt 2 len 4 (sig created 2021-06-27)
	hashed subpkt 27 len 1 (key flags: 0C)
	hashed subpkt 9 len 4 (key expires after 2y0d0h0m)
	subpkt 16 len 8 (issuer key ID C3C828AFB6681C17)
	data: [2047 bits]
```

Note that values are the same, except the new key has a `new-ctb` tag, whatever that means. That causes the key encoding to be completely different.

This impacts searching for PGP keys, and the feature I'm trying to add to return a set of verifiers for each artifact type (which would include the key value which should match the input)

**Comment by loosebazooka** [comment]
@adamkorcz looking into fuzzing here.

**Comment by Hayden-IO** [comment]
note you’ll mostly be testing the x/crypto/openpgp library which is deprecated. If we find anything critical, we will need to maintain a fork if upstream won’t accept changes.

**Comment by Hayden-IO** [comment]
Chatted offline about this. For one, there really is nothing we can do due to library limitations. Two, we also need to canonical the input. We should document that verifiers should search based on fingerprint rather than exact value.
